jscomp/others/js_typed_array.ml

```diff
@@ -164,11 +161,14 @@ end
 
 
 (**/**)
-module TypedArray (Type: Type) : S with type elt = Type.t  = struct
+module TypedArray (Type: Type) : S with
+  type elt = Type.t
+  and type 'a typed_array = 'a Js_typed_array2.TypedArray(Type).typed_array
```

Comment:
@bobzhang I had to un-abstract the abstract types here, because this functor is used below to create the concrete modules, e.g. `Int16Array`. If `Int16Array.typed_array` stayed abstract, then this'd be a breaking change, since the `Int16Array` from this `Js.TypedArray` file wouldn't interoperate with the `Int16Array` from the new `Js.TypedArray2` file.

Is this the right way of doing it?

Reply:
This makes things complicated, can we do some dirty work but provide users an easier user interface, my ideas, you may need refine it:
```
#define MAKEBASIC(elt)  \
  type 'a typed_array \
  type t = elt typed_array \
  external unsafe_get : t -> int -> elt  = "" [@@bs.get_index] \
  external unsafe_set : t -> int -> elt -> unit = "" [@@bs.set_index] \
  external buffer : t -> array_buffer = "" [@@bs.get] \
  external byteLength : t -> int = "" [@@bs.get] \
  external byteOffset : t -> int = "" [@@bs.get] \
  external setArray : elt array -> unit = "set" [@@bs.send.pipe: t] \
  external setArrayOffset : elt array -> int -> unit = "set" [@@bs.send.pipe: t] \
  (* There's also an overload for typed arrays, but don't know how to model that without subtyping *) \
  external length : t -> int = "" [@@bs.get] \
  (* Mutator functions *) \
  external copyWithin : to_:int -> t = "" [@@bs.send.pipe: t] \
  external copyWithinFrom : to_:int -> from:int -> t = "copyWithin" [@@bs.send.pipe: t]\
  external copyWithinFromRange : to_:int -> start:int -> end_:int -> t = "copyWithin" [@@bs.send.pipe: t]\

module  IntArray = struct 
  MAKEBASIC(int)  
end  

module FloatArray = struct 
  MAKEBASIC(float)
end 
```